include/pybind11/detail/internals.h

```diff
@@ -18,6 +18,27 @@ inline PyTypeObject *make_static_property_type();
 inline PyTypeObject *make_default_metaclass();
 inline PyObject *make_object_base_type(PyTypeObject *metaclass);
 
+// The old Python Thread Local Storage (TLS) API is deprecated in Python 3.7 in favor of the new
+// Thread Specific Storage (TSS) API.
+#if PY_VERSION_HEX >= 0x03070000
+    #define PYBIND11_TLS_KEY_TYPE Py_tss_t *
+    #define PYBIND11_TLS_KEY_NULL_VALUE nullptr
```

Comment:
Do the type and null value need to be separate here?  It looks from the way they are used that you could simplify it to something along the lines of:

```C++
#define PYBIND11_TLS_KEY_INIT(var) Py_tss_t *var = nullptr;
// or for <3.7:
#define PYBIND11_TLS_KEY_INIT(var) decltype(PyThread_create_key()) var = 0;
```

Reply:
Hmmm, myeah, true. Depends if at some point you'd need to have the type separate.

But I can still change that, if you want? (Or feel free to push that commit to this branch yourself, as you should be allowed by GitHub.)

Reply:
Considering that these are entirely internal macros (i.e. if you're using them outside of pybind11, shame on you) we can also break them up later if needed -- but for the moment it doesn't look like they are.

Reply:
(Pushed a commit)